### PR TITLE
Support array declarations

### PR DESCRIPTION
**Pull request by woodrowbarlow**

### CODE REVIEW

File: jeff65/gold/lexer.py [hunk 1]
```diff
@@ -95,7 +96,10 @@
 }
 
 # non-whitespace characters which can end words.
-specials = '()[]{}:.,"\\'
+specials = '()[]{}:;.,"\\@'
+
+# non-whitespace characters which can end sprinkles.
+sprinkle_specials = ','
```
Comment (woodrowbarlow) [inline jeff65/gold/lexer.py]: TODO: this is a hack and doesn't even fix the nested brackets problem

File: tests/test_ast.py [hunk 2]
```diff
@@ -149,6 +149,120 @@ def test_let_without_storage_class():
     assert_equal("u8", b.lhs.rhs.text)
 
 
+def test_array_declaration():
```
Comment (woodrowbarlow) [inline tests/test_ast.py]: TODO: these tests are long and unreadable. see what can be done about that.

File: tests/test_ast.py [hunk 3]
```diff
@@ -149,6 +149,120 @@ def test_let_without_storage_class():
...
+    assert_is_instance(v.lhs, ast.NumericNode)
+    assert_equal(v.lhs.text, "0")
+    assert_is_instance(v.rhs, ast.PunctuationCommaNode)
+    assert_is_instance(v.rhs.lhs, ast.NumericNode)
+    assert_equal(v.rhs.lhs.text, "1")
+    assert_is_instance(v.rhs.rhs, ast.NumericNode)
+    assert_equal(v.rhs.rhs.text, "2")
+
+
+def test_array_multidiminsional():
```
Comment (woodrowbarlow) [inline tests/test_ast.py]: TODO: this test is failing because the lexer is taking "[[" and "]]" as single tokens.

File: jeff65/gold/lexer.py (hunk 1 above)
Comment (woodrowbarlow) [reply jeff65/gold/lexer.py]: it does prevent `],` from being read as a single token, but so would moving `[` and `]` to the `nestable_delimiters` list. that still won't help disambiguate between the comment close and two `]` tokens, though.

File: jeff65/gold/lexer.py [hunk 4]
```diff
@@ -210,13 +212,18 @@ def lex(stream):
...
         # nestable delimiters (parentheses)
-        elif c in nestable_delimiters:
+        elif comment_depth is 0 and c in nestable_delimiters:
             yield make_token(position, c)
 
         # non-alphanumeric word (mostly operators)
         # we call these "sprinkles"
         else:
-            yield make_token(position, scan_sprinkle(source, c))
+            t = make_token(position, scan_sprinkle(source, c))
```
Comment (woodrowbarlow) [inline jeff65/gold/lexer.py]: keeping track of comment nesting depth in the lexer doesn't seem like an optimal solution, but it's the best one that has occurred to me so far.

@jdpage do you have any ideas? i would really rather not use parentheses on the right-hand side of an array declaration.

File: tests/test_ast.py (hunk 2 above)
Comment (woodrowbarlow) [reply tests/test_ast.py]: done.

File: jeff65/gold/lexer.py (hunk 4 above)
Comment (woodrowbarlow) [reply jeff65/gold/lexer.py]: actually, we have another problem. say i have:

    let x: [u8; 2, 2] = [[0, 1], [2, 3]]

and i want to comment it out:

    --[[ let x: [u8; 2, 2] = [[0, 1], [2, 3]] ]]

that's a problem.

i think we have no choice except to either use parentheses on the right of the assignment or to change the syntax for comments.

i don't like parentheses on the right because parentheses are already used for multiple things.

Comment (woodrowbarlow) [reply jeff65/gold/lexer.py]: i know you think they're ugly, but now we have a perfect example of why so many languages use `/*` and `*/`. we need a better argument than "they're ugly".

Comment (jdpage) [reply jeff65/gold/lexer.py]: I actually don't see the problem; `--[[x=values[offsets[y]]]]` is unambiguous. Just don't think of the `]]` comment-ender as a single token; it's really two tokens, and the comment-opener is looking for two consecutive `]` tokens to match instead of a single `]]` token.

Comment (woodrowbarlow) [reply jeff65/gold/lexer.py]: i am assuming comments should follow these syntax rules (from the user point of view):

1. a comment can contain any text without breaking the parser, except for the "open comment" and "end comment" token.
2. there must be exactly one "close comment" token for each "open comment" token. this supports nesting.

let's consider the naive implementation of parsing `--[[x=values[offsets[y]]]]`. it would result in a comment containing `x=values[offsets[y` and two unmatched close brackets left in the stream (because the comment ends at the first instance of double-close-bracket). this is obviously wrong.

a "smarter" way of parsing that would be:

1. the `--` token creates a `CommentNode`. all of the square brackets would just create square bracket nodes (open or close as appropriate).
2. the `eat_comment` function of `CommentNode` would:
   1. make sure that the next two tokens in the stream are open bracket nodes
   2. set `depth` to 2.
   3. eat right tokens until depth is zero. each open bracket increments depth, each close bracket decrements depth.
   4. do something to make sure that the last two brackets were adjacent.

these rules would properly parse `--[[x=values[offsets[y]]]]`; however, it violates rule 1. i can no longer put unmatched brackets inside of a comment. this is unexpected and violates a pretty core concept of "comments". i consider that unacceptable.

Comment (woodrowbarlow) [reply jeff65/gold/lexer.py]: another option is using bare strings as comments, like in python.

**Comment by woodrowbarlow** [comment]
that last commit is just to see what the code would look like if we were using `/*` and `*/` for comments.

i don't feel strongly about `/*` in particular, but if you agree with my long comment above, i do believe that changing the token to anything that is otherwise an illegal character combination would be the best solution.

i'd also be happy with bare-strings-as-comments, but that would mean we can only put comments at statement-level (rather than inside of expressions). i don't really consider that a disadvantage.

**Comment by woodrowbarlow** [comment]
bare strings as comments wouldn't let us arbitrarily nest comments like we want to because the begin and end delimiters aren't distinct. sure, we could implement python's "triple string", which would allow you to nest single strings inside them without escaping the quotes, but you can't nest a triple string inside another triple string. that's a relatively minor sacrifice, but it seems wrong to be making any sacrifices at all just because `/*` is "ugly".

basically, i think you need to make a better argument against `/*` other than just aesthetics. to be clear, i do see the value of aesthetics, but avoiding language limitations and unexpected rules will always win out over aesthetics.
